kernel/src/engine/default/json.rs

```diff
@@ -127,90 +179,309 @@ pub struct JsonOpener {
+                wakers: HashMap::new(),
+            };
+
+            Self {
+                inner,
+                state: Mutex::new(state),
+            }
+        }
+    }
+
+    impl<T: ObjectStore> std::fmt::Display for OrderedGetStore<T> {
+        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+            let state = self.state.lock().unwrap();
+            write!(f, "OrderedGetStore({:?})", state.ordered_keys)
+        }
+    }
+
+    #[async_trait::async_trait]
+    impl<T: ObjectStore> ObjectStore for OrderedGetStore<T> {
+        async fn put(&self, location: &Path, payload: PutPayload) -> Result<PutResult> {
+            self.inner.put(location, payload).await
+        }
+
+        async fn put_opts(
+            &self,
+            location: &Path,
+            payload: PutPayload,
+            opts: PutOptions,
+        ) -> Result<PutResult> {
+            self.inner.put_opts(location, payload, opts).await
+        }
+
+        async fn put_multipart(&self, location: &Path) -> Result<Box<dyn MultipartUpload>> {
+            self.inner.put_multipart(location).await
+        }
+
+        async fn put_multipart_opts(
+            &self,
+            location: &Path,
+            opts: PutMultipartOpts,
+        ) -> Result<Box<dyn MultipartUpload>> {
+            self.inner.put_multipart_opts(location, opts).await
+        }
+
+        // A GET request is fulfilled by checking if the requested path is next in order:
+        // - if yes, remove the path from the queue and proceed with the GET request, then wake the
+        //   next path in order
+        // - if no, register the waker and wait
+        async fn get(&self, location: &Path) -> Result<GetResult> {
+            // Do the actual GET request first, then introduce any artificial ordering delays as needed
+            let result = self.inner.get(location).await;
+
+            // we implement a future which only resolves once the requested path is next in order
+            future::poll_fn(move |cx| {
+                let mut state = self.state.lock().unwrap();
+                let Some(next_key) = state.ordered_keys.front() else {
+                    panic!("Ran out of keys before {location}");
+                };
+                if next_key == location {
+                    // We are next in line. Nobody else can remove our key, and our successor
+                    // cannot race with us to register itself because we hold the lock.
+                    //
+                    // first, remove our key from the queue.
+                    //
+                    // note: safe to unwrap because we just checked that the front key exists (and
+                    // is the same as our requested location)
+                    state.ordered_keys.pop_front().unwrap();
+
+                    // there are three possible cases, either:
+                    // 1. the next key has a waker already registered, in which case we wake it up
```

Comment:
nit: maybe note that this is the case where something has already requested the next key in line, so that's why there is a waker waiting, and we need to wake it up

Reply:
added more!